{{ cookiecutter.package_name|slugify }}/{% if cookiecutter.continuous_integration == "GitLab" %}.gitlab-ci.yml{% endif %}

```diff
@@ -13,48 +22,49 @@ stages:
-        if [ -n "$CI_COMMIT_TAG" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_COMMIT_TAG"; fi
-        if [ -n "$CI_ENVIRONMENT_NAME" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_ENVIRONMENT_NAME"; fi
-        DOCKER_TAGS_JOINED=""
-        for DOCKER_TAG in $DOCKER_TAGS; do
-          DOCKER_TAGS_JOINED="$DOCKER_TAGS_JOINED --tag $DOCKER_IMAGE:$DOCKER_TAG"
-        done
 
-        # Build the Docker image with all of the selected tags.
+      # Compile a list of image tags consisting of a hash of its contents, the latest tag if this
+      # pipeline is running for the default branch, and the Git tag if this commit is tagged.
+      DOCKER_TAGS="$DOCKER_IMAGE_SHA"
+      if [ "$CI_COMMIT_BRANCH" = "$CI_DEFAULT_BRANCH" ]; then DOCKER_TAGS="$DOCKER_TAGS latest"; fi
+      if [ -n "$CI_COMMIT_TAG" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_COMMIT_TAG"; fi
+      if [ -n "$CI_ENVIRONMENT_NAME" ]; then DOCKER_TAGS="$DOCKER_TAGS $CI_ENVIRONMENT_NAME"; fi
+      DOCKER_TAGS_JOINED=""
+      for DOCKER_TAG in $DOCKER_TAGS; do
+        DOCKER_TAGS_JOINED="$DOCKER_TAGS_JOINED --tag $DOCKER_IMAGE:$DOCKER_TAG"
+      done
+
+      # Build the Docker image with all of the selected tags.
+      {%- if cookiecutter.private_package_repository_name %}
+      echo "[http-basic.{{ cookiecutter.private_package_repository_name|slugify }}]" >> auth.toml
+      echo "username = \"gitlab-ci-token\"" >> auth.toml
+      echo "password = \"$CI_JOB_TOKEN\"" >> auth.toml
+      {%- endif %}
+      DOCKER_BUILDKIT=1 docker build \
+        --cache-from "${DOCKER_BASE_IMAGE:-ci}" \
+        --build-arg SOURCE_BRANCH="$CI_COMMIT_REF_NAME" \
+        --build-arg SOURCE_COMMIT="$CI_COMMIT_SHA" \
+        --build-arg SOURCE_TIMESTAMP="$CI_COMMIT_TIMESTAMP" \
         {%- if cookiecutter.private_package_repository_name %}
-        echo "[http-basic.{{ cookiecutter.private_package_repository_name|slugify }}]" >> auth.toml
-        echo "username = \"gitlab-ci-token\"" >> auth.toml
-        echo "password = \"$CI_JOB_TOKEN\"" >> auth.toml
+        --secret id=poetry_auth,src=auth.toml \
         {%- endif %}
-        DOCKER_BUILDKIT=1 docker build \
-          --build-arg SOURCE_BRANCH="$CI_COMMIT_REF_NAME" \
-          --build-arg SOURCE_COMMIT="$CI_COMMIT_SHA" \
-          --build-arg SOURCE_TIMESTAMP="$CI_COMMIT_TIMESTAMP" \
-          {%- if cookiecutter.private_package_repository_name %}
-          --secret id=poetry_auth,src=auth.toml \
-          {%- endif %}
-          --target "$DOCKER_TARGET" \
-          --pull \
-          $DOCKER_TAGS_JOINED \
-          .
+        --target "$DOCKER_TARGET" \
+        --pull \
+        $DOCKER_TAGS_JOINED \
+        .
 
-        # Push all the tagged images.
-        for DOCKER_TAG in $DOCKER_TAGS; do
-          docker push "$DOCKER_IMAGE:$DOCKER_TAG"
-        done
-      fi
+      # Push all the tagged images.
+      for DOCKER_TAG in $DOCKER_TAGS; do
+        docker push "$DOCKER_IMAGE:$DOCKER_TAG"
+      done
+  rules:
+    - changes:
```

Comment:
Do I understand correctly that only changes to the files listed below _within the active branch_ will trigger this job? What happens on the first commit of a new branch?

Reply:
Here's the docs:

https://docs.gitlab.com/ee/ci/yaml/#ruleschanges

The `rules:changes` conditions will be checked every time a push happens. First commit of a new branch should therefore not rerun the Docker build. However, the docs indicate that things like scheduled runs, manual runs, tag builds, ... will ignore `rules:changes`. I'll re-add the check for an existing `ci` image, but leaving the `rules:changes` in will still give large time savings (since the job doesn't even need to be executed).